File: tianshu/core/rollback.py
from __future__ import annotations

import shutil
import time
from pathlib import Path

from tianshu.config import SENSITIVE_DIR as SN_DIR
from tianshu.config import WORKSPACE_DIR as WS_DIR

SNAP_ROOT = WS_DIR / ".ts-snapshots"
MAX_SNAPSHOTS = 20

_SKIP_DIRS = {SNAP_ROOT.name, SN_DIR.name, ".git"}


def _safe_label(label: str) -> str:
    return "".join(c if c.isalnum() or c in "-_." else "_" for c in label)[:40]


def _new_snap_dir(label: str = "") -> Path:
    stamp = time.strftime("%Y%m%d-%H%M%S")
    SNAP_ROOT.mkdir(parents=True, exist_ok=True)
    base = f"snap-{stamp}"
    if label:
        base = f"{base}-{_safe_label(label)}"
    d = SNAP_ROOT / base
    n = 2
    while d.exists():
        d = SNAP_ROOT / f"{base}-{n}"
        n += 1
    d.mkdir()
    return d


def _prune() -> None:
    snaps = sorted(SNAP_ROOT.glob("snap-*"), key=lambda p: p.name)
    for old in snaps[:-MAX_SNAPSHOTS]:
        shutil.rmtree(old, ignore_errors=True)


def auto_snapshot(path: str | Path) -> str | None:
    p = Path(path).resolve()
    wp = WS_DIR.resolve()
    try:
        rel = p.relative_to(wp)
    except ValueError:
        return None
    if not p.is_file():
        return None
    snap = _new_snap_dir("auto")
    dst = snap / rel
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(p, dst)
    _prune()
    return snap.name


def snapshot_all(label: str = "manual") -> str:
    wp = WS_DIR.resolve()
    snap = _new_snap_dir(label)
    n = 0
    for p in wp.rglob("*"):
        if not p.is_file():
            continue
        rel = p.relative_to(wp)
        if any(part in _SKIP_DIRS for part in rel.parts):
            continue
        dst = snap / rel
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(p, dst)
        n += 1
    _prune()
    return f"{snap.name}({n} 个文件)"


def list_snapshots(limit: int = 10) -> str:
    snaps = sorted(SNAP_ROOT.glob("snap-*"), key=lambda p: p.name, reverse=True)
    if not snaps:
        return "(暂无快照)"
    lines = []
    for d in snaps[:limit]:
        count = sum(1 for f in d.rglob("*") if f.is_file())
        lines.append(f"{d.name}  |  {count} 个文件")
    return "\n".join(lines)


def restore_snapshot(snapshot: str, target: str) -> str:
    wp = WS_DIR.resolve()
    snap = SNAP_ROOT / snapshot
    if not snap.is_dir():
        raise FileNotFoundError(f"快照不存在: {snapshot}")
    t = Path(target)
    if not t.is_absolute():
        t = wp / t
    t = t.resolve()
    try:
        rel = t.relative_to(wp)
    except ValueError:
        raise PermissionError(f"目标必须在工作区内: {target}")
    src = snap / rel
    if not src.exists():
        raise FileNotFoundError(f"快照中没有该路径: {rel}")
    auto_snapshot(t)
    t.parent.mkdir(parents=True, exist_ok=True)
    if src.is_dir():
        shutil.copytree(src, t, dirs_exist_ok=True)
        return f"已从 {snap.name} 恢复目录 {rel}(恢复前版本已自动备份)"
    shutil.copy2(src, t)
    return f"已从 {snap.name} 恢复文件 {rel}(恢复前版本已自动备份)"

File: tianshu-ts/src/core/rollback.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { SENSITIVE_DIR, WORKSPACE_DIR } from "../config.js";
import { audit } from "./audit.js";

const SNAP_ROOT = path.join(WORKSPACE_DIR, ".ts-snapshots");
const MAX_SNAPSHOTS = 20;
const SKIP_DIRS = new Set([".ts-snapshots", ".ts-secrets", ".git"]);

function safeLabel(label: string): string {
  return label.replace(/[^a-zA-Z0-9\-_.]/g, "_").slice(0, 40);
}

function newSnapDir(label = ""): string {
  const stamp = new Date().toISOString().slice(0, 19).replace(/[-:T]/g, (ch) => (ch === "T" ? "-" : ch)).replace(/-/g, "");
  let base = `snap-${stamp}`;
  if (label) base = `${base}-${safeLabel(label)}`;
  fs.mkdirSync(SNAP_ROOT, { recursive: true });
  let d = path.join(SNAP_ROOT, base);
  let n = 2;
  while (fs.existsSync(d)) {
    d = path.join(SNAP_ROOT, `${base}-${n}`);
    n++;
  }
  fs.mkdirSync(d);
  return d;
}

function prune(): void {
  let snaps: string[];
  try {
    snaps = fs.readdirSync(SNAP_ROOT).filter((n) => n.startsWith("snap-")).sort();
  } catch {
    return;
  }
  for (const old of snaps.slice(0, Math.max(0, snaps.length - MAX_SNAPSHOTS))) {
    fs.rmSync(path.join(SNAP_ROOT, old), { recursive: true, force: true });
  }
}

export function autoSnapshot(filePath: string): string | null {
  const p = path.resolve(filePath);
  const wp = path.resolve(WORKSPACE_DIR);
  if (!p.startsWith(wp + path.sep)) return null;
  if (!fs.existsSync(p) || !fs.statSync(p).isFile()) return null;
  const snap = newSnapDir("auto");
  const rel = path.relative(wp, p);
  const dst = path.join(snap, rel);
  fs.mkdirSync(path.dirname(dst), { recursive: true });
  fs.copyFileSync(p, dst);
  prune();
  return path.basename(snap);
}

export function snapshotAll(label = "manual"): string {
  const wp = path.resolve(WORKSPACE_DIR);
  const snap = newSnapDir(label);
  let n = 0;
  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      const full = path.join(dir, e.name);
      if (e.isDirectory()) {
        if (!SKIP_DIRS.has(e.name)) walk(full);
      } else {
        const rel = path.relative(wp, full);
        const dst = path.join(snap, rel);
        fs.mkdirSync(path.dirname(dst), { recursive: true });
        fs.copyFileSync(full, dst);
        n++;
      }
    }
  };
  walk(wp);
  prune();
  return `${path.basename(snap)}(${n} 个文件)`;
}

export function listSnapshots(limit = 10): string {
  let snaps: string[];
  try {
    snaps = fs.readdirSync(SNAP_ROOT).filter((n) => n.startsWith("snap-")).sort().reverse();
  } catch {
    return "(暂无快照)";
  }
  if (!snaps.length) return "(暂无快照)";
  const lines = snaps.slice(0, limit).map((name) => {
    let count = 0;
    const walk = (dir: string): void => {
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      for (const e of entries) {
        const full = path.join(dir, e.name);
        if (e.isDirectory()) walk(full);
        else count++;
      }
    };
    walk(path.join(SNAP_ROOT, name));
    return `${name}  |  ${count} 个文件`;
  });
  return lines.join("\n");
}

export function restoreSnapshot(snapshot: string, target: string): string {
  const wp = path.resolve(WORKSPACE_DIR);
  const snap = path.join(SNAP_ROOT, snapshot);
  if (!fs.existsSync(snap) || !fs.statSync(snap).isDirectory()) {
    throw new Error(`快照不存在: ${snapshot}`);
  }
  const t = path.isAbsolute(target) ? path.resolve(target) : path.resolve(wp, target);
  if (!t.startsWith(wp + path.sep)) {
    throw new Error(`目标必须在工作区内: ${target}`);
  }
  const rel = path.relative(wp, t);
  const src = path.join(snap, rel);
  if (!fs.existsSync(src)) {
    throw new Error(`快照中没有该路径: ${rel}`);
  }
  autoSnapshot(t);
  if (fs.statSync(src).isDirectory()) {
    fs.mkdirSync(t, { recursive: true });
    fs.cpSync(src, t, { recursive: true });
  } else {
    fs.mkdirSync(path.dirname(t), { recursive: true });
    fs.copyFileSync(src, t);
  }
  audit("rollback.restore", `snapshot=${snap} target=${rel}`);
  return `已从 ${snapshot} 恢复 ${rel}(恢复前版本已自动备份)`;
}